src/test/rpc/NoRippleCheck_test.cpp

```diff
@@ -254,27 +254,43 @@ class NoRippleCheckLimits_test : public beast::unit_test::suite
         env (fset (alice, asfDefaultRipple));
         env.close ();
 
-        for (auto i = 0; i < ripple::RPC::Tuning::noRippleCheck.rmax + 5; ++i)
+        // keep track of the last endoint balance so that we know
+        // when we are approaching the dropThreshold limit.
+        // only used for non-admin case (see below)
+        int lastBalance{0};
+        auto checkBalance = [&lastBalance, &env]()
         {
-            if (! admin)
+            // this is endpoint drop prevention. Non admin ports will drop
+            // requests if they are coming too fast, so we manipulate the
+            // resource manager here to reset the enpoint balance (for
+            // localhost) if we get too close to the drop limit. It would
+            // be better if we could add this functionality to Env somehow
+            // or otherwise disable endpoint charging for certain test
+            // cases.
+            using namespace ripple::Resource;
+            using namespace std::chrono;
+            using namespace beast::IP;
+            auto c = env.app().getResourceManager()
+                .newInboundEndpoint (Endpoint::from_string (test::getEnvLocalhostAddr()));
+
+            // if we are within 4x the current balance delta of the
+            // threshold, go ahead and reset the balance. The goal here is
+            // to prevent connection drops and 4x is just a heuristic to
+            // know when we are approaching the limit.
+            if (dropThreshold - c.balance() <= 4*(c.balance() - lastBalance))
```

Comment:
Occasionally c.balance() is lower than lastBalance, a negative value.

An alternative is to check if we at warning level. 
eg.
```
        auto checkBalance = [&env]()
        {
            using namespace ripple::Resource;
            using namespace std::chrono;
            auto c {env.app().getResourceManager()
                .newInboundEndpoint(Endpoint::from_string(test::getEnvLocalhostAddr()))};
            if (c.balance() > warningThreshold)
            {
                using clock_type = beast::abstract_clock <steady_clock>;
                c.entry().local_balance =
                    DecayingSample <decayWindowSeconds, clock_type>
                { steady_clock::now() };
            }
        };
```
I think a better solution is to remove the lambda altogether and use the following before the loop.
```
        if (!admin)
        {
            // Prevent non admin ports from dropping requests
            using namespace ripple::Resource;
            using namespace std::chrono;
            using namespace beast::IP;
            auto c {env.app().getResourceManager().newInboundEndpoint(
                Endpoint::from_string(test::getEnvLocalhostAddr()))};
            c.entry().remote_balance = -dropThreshold;
        }

        for (auto i = 0; i < ripple::RPC::Tuning::noRippleCheck.rmax + 5; ++i)
        {
            auto& txq = env.app().getTxQ();
            auto const gw = Account {"gw" + std::to_string(i)};
            ...
```